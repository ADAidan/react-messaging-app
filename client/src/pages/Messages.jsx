/* eslint-disable no-underscore-dangle */
import * as React from "react";
import axios from "axios";
import Container from "@mui/material/Container";
import Typography from "@mui/material/Typography";
import Paper from "@mui/material/Paper";
import { Box, IconButton, Stack, Toolbar } from "@mui/material";
import AddIcon from "@mui/icons-material/Add";
import { useNavigate } from "react-router-dom";
import socket from "../socket";
import { Message, MessageContent } from "../components/Message";
import MessageInput from "../components/MessageInput";
import ChatCard from "../components/ChatCard";
import NoMessages from "../components/NoMessages";
import NoDirectMessages from "../components/NoDirectMessages";
import AddConversationModal from "../components/AddConversationModal";
import UserContext from "../Context";

const formatTime = (isoString) => {
  const date = new Date(isoString);

  const options = {
    hour: "numeric",
    minute: "numeric",
    hour12: true,
  };

  return date.toLocaleTimeString("en-US", options);
};

const bundleMessages = (messages) => {
  const combinedMessages = [];

  let currentMessage = {};

  messages.forEach((message, index) => {
    if (
      currentMessage.header &&
      currentMessage.header.author === message.author.username
    ) {
      currentMessage.messagesContent.push({
        id: message._id,
        content: message.content,
        timeSent: message.formattedTime,
      });
    } else {
      if (currentMessage.header) {
        combinedMessages.push(currentMessage);
      }

      // create a new current message object
      currentMessage = {
        id: combinedMessages.length,
        header: {
          profilePicture: message.author.profilePicture,
          author: message.author.username,
          time: message.formattedTime,
        },
        messagesContent: [
          {
            id: message._id,
            content: message.content,
            timeSent: message.formattedTime,
          },
        ],
      };
    }
    if (index === messages.length - 1) {
      combinedMessages.push(currentMessage);
    }
  });

  return combinedMessages;
};

function Messages() {
  const userId = sessionStorage.getItem("user");
  const userContextData = React.useContext(UserContext);
  const messageContainerRef = React.useRef(null);
  const [directMessages, setDirectMessages] = React.useState([]);
  const [displayedMessages, setDisplayedMessages] = React.useState([]);
  const [selectedChat, setSelectedChat] = React.useState(null);
  const [open, setOpen] = React.useState(false); // AddModal open state
  const [userContacts, setUserContacts] = React.useState([]); // contacts to display on the AddModal

  const navigate = useNavigate();

  // If user is not logged in, redirect to login page
  React.useEffect(() => {
    axios
      .get(`${import.meta.env.VITE_API_URL}/users/protected`, {
        withCredentials: true,
      })
      .then(() => {
        // eslint-disable-next-line no-console
        console.log("accessed protected route");
      })
      .catch(() => {
        navigate("/logout");
      });
  }, []);

  // Join chat room
  const handleJoinChat = (chatId) => {
    setSelectedChat(chatId);
    socket.emit("JoinRoom", chatId);
  };

  const getContactData = (participants) => {
    const contact = participants.find(
      (participant) => participant._id !== userId,
    );
    if (!contact.username && !contact.profilePicture) {
      return {
        contactName: "Unknown",
        contactProfilePicture: "",
        contactStatus: "Unknown",
      };
    }
    return {
      contactName: contact.username,
      contactProfilePicture: contact.profilePicture,
      contactStatus: contact.status,
      contactId: contact._id,
    };
  };

  // Socket event listener for adding a new conversation
  React.useEffect(() => {
    const handleNewConversation = async (conversation) => {
      const contactId = conversation.participants.find(
        (participant) => participant !== userId,
      );
      if (!contactId) return;
      try {
        const response = await axios.get(
          `${import.meta.env.VITE_API_URL}/users/${contactId}/info`, // gets the contacts info for the chat card
        );
        const userData = response.data;

        const newChat = {
          _id: conversation._id,
          username: userData.username,
          status: userData.status,
          profilePicture: userData.profilePicture,
          messages: [],
        };

        setDirectMessages((prevChats) => [...prevChats, newChat]);
      } catch (error) {
        // eslint-disable-next-line no-console
        console.error("Error fetching conversation:", error);
      }
    };
    socket.on("addConversation", handleNewConversation);
    return () => {
      socket.off("addConversation");
    };
  }, [directMessages]);

  const lastMessageRef = React.useRef(null);

  React.useEffect(() => {
    if (displayedMessages.length > 0) {
      lastMessageRef.current = displayedMessages[displayedMessages.length - 1];
    }
  }, [displayedMessages]);

  // Socket event listener for creating a new message
  React.useEffect(() => {
    socket.on("receiveMessage", (message) => {
      axios
        .get(`${import.meta.env.VITE_API_URL}/users/${userId}`)
        .then((response) => {
          const user = response.data;

          const lastMessage = lastMessageRef.current;

          const formattedTime = formatTime(message.createdAt);

          const newMessage = {
            id: displayedMessages.length,
            header: {
              profilePicture: user.profilePicture,
              author: user.username,
              time: formattedTime,
            },
            messagesContent: [
              {
                id: message._id,
                content: message.content,
                timeSent: formattedTime,
              },
            ],
          };

          if (lastMessage && lastMessage.header.author === user.username) {
            lastMessage.messagesContent.push({
              id: message._id,
              content: message.content,
              timeSent: formattedTime,
            });
            setDisplayedMessages((prevMessages) => [...prevMessages]);
          } else {
            setDisplayedMessages((prevMessages) => [
              ...prevMessages,
              newMessage,
            ]);
          }
        });
    });
    return () => {
      socket.off("receiveMessage");
    };
  }, []);

  // Socket event listener for deleting a conversation
  React.useEffect(() => {
    socket.on("DeleteConversation", (conversationId) => {
      setDirectMessages((prevChats) =>
        prevChats.filter((chat) => chat._id !== conversationId),
      );
    });
    return () => {
      socket.off("DeleteConversation");
    };
  }, [directMessages]);

  React.useEffect(() => {
    if (!userId) return;
    socket.emit("JoinRoom", userId);
    const fetchUserContacts = async () => {
      try {
        const response = await axios.get(
          `${import.meta.env.VITE_API_URL}/users/${userId}/contacts`,
        );
        const updatedContacts = response.data.map((contact) => ({
          ...contact,
          isContact: true,
        }));
        setUserContacts(updatedContacts);
      } catch (error) {
        // eslint-disable-next-line no-console
        console.error("Axios Error:", error);
      }
    };
    fetchUserContacts();
  }, [userId]);

  React.useEffect(() => {
    if (!messageContainerRef.current) return;
    messageContainerRef.current.scrollTop =
      messageContainerRef.current.scrollHeight;
  }, [displayedMessages]);

  React.useEffect(() => {
    if (!userId) return;
    // eslint-disable-next-line consistent-return
    const getDirectMessages = async () => {
      try {
        const response = await axios.get(
          `${import.meta.env.VITE_API_URL}/users/${userId}/direct-messages`,
        );
        const conversationData = response.data;
        const conversations = conversationData.map((conversation) => {
          const {
            contactName,
            contactProfilePicture,
            contactStatus,
            contactId,
          } = getContactData(conversation.participants);
          return {
            _id: conversation._id,
            contactId,
            username: contactName,
            status: contactStatus,
            profilePicture: contactProfilePicture,
            messages: conversation.messages,
          };
        });
        // conversationData needs to be an array of objects with username, status and profilePicture properties
        setDirectMessages(conversations);
      } catch (error) {
        // eslint-disable-next-line no-console
        console.error("Axios Error:", error);
      }
    };
    getDirectMessages();
  }, []);

  React.useEffect(() => {
    if (!directMessages.length) return;
    if (!selectedChat) {
      handleJoinChat(directMessages[0]._id);
      return;
    }
    const getDisplayedMessages = () => {
      const currentChat = directMessages.find(
        (chat) => chat._id === selectedChat,
      );
      if (!currentChat) return [];
      return currentChat.messages;
    };
    const messagesData = getDisplayedMessages();

    const formattedMessages = messagesData.map((message) => ({
      _id: message._id,
      author: {
        username: message.author.username,
        profilePicture: message.author.profilePicture,
      },
      content: message.content,
      formattedTime: formatTime(message.createdAt),
    }));

    const bundledMessages = bundleMessages(formattedMessages);

    setDisplayedMessages(bundledMessages);
  }, [directMessages, selectedChat]);

  const handleMouseDownAdd = (e) => {
    e.preventDefault();
  };

  const handleClickAddChat = () => {
    setOpen(true);
  };

  return (
    <Container
      maxWidth="lg"
      disableGutters
      sx={{
        height: "100vh",
        my: 0,
      }}
    >
      <Stack
        direction="column"
        spacing={0}
        sx={{
          height: "100vh",
        }}
      >
        <Toolbar />
        <Stack
          direction="row"
          spacing={0}
          sx={{
            height: "calc(100vh - 64px)",
          }}
        >
          <Paper
            elevation={1}
            sx={{
              px: 3,
              my: 0,
              overflowY: "auto",
              borderRadius: 0,
              flex: 1,
            }}
          >
            <Stack
              direction="row"
              spacing={2}
              sx={{
                p: 1,
                justifyContent: "space-between",
                alignItems: "center",
              }}
            >
              <Typography variant="h5" component="h1" sx={{ mb: 2 }}>
                Direct Message
              </Typography>
              <IconButton
                aria-label="add chat"
                onClick={handleClickAddChat}
                onMouseDown={handleMouseDownAdd}
              >
                <AddIcon />
              </IconButton>
              <AddConversationModal
                open={open}
                setOpen={setOpen}
                setSelectedChat={setSelectedChat}
                allUsers={userContacts}
              />
            </Stack>
            {directMessages.length > 0 &&
              directMessages.map((chat) => (
                <ChatCard
                  key={chat._id}
                  chat={chat}
                  handleJoinChat={handleJoinChat}
                  selected={selectedChat === chat._id}
                />
              ))}
          </Paper>
          {selectedChat ? (
            <Paper
              elevation={1}
              sx={{
                position: "relative",
                borderRadius: 0,
                flex: 2,
              }}
            >
              <Stack
                spacing={0}
                sx={{
                  height: "100%",
                }}
              >
                <Stack
                  data-testid="messages-container"
                  ref={messageContainerRef}
                  spacing={2}
                  sx={{
                    overflowY: "auto",
                    p: 1,
                  }}
                >
                  {displayedMessages.length > 0 ? (
                    displayedMessages.map((message) => (
                      <Message key={message.id} messageHeader={message.header}>
                        {message.messagesContent.map((content) => (
                          <MessageContent
                            key={content.id}
                            content={content.content}
                            bgColor={
                              userContextData.username === message.header.author
                                ? "#f3f3f3"
                                : "#1e90ff"
                            }
                          />
                        ))}
                      </Message>
                    ))
                  ) : (
                    <NoMessages />
                  )}
                </Stack>
                <MessageInput selectedChat={selectedChat} />
              </Stack>
            </Paper>
          ) : (
            <Box
              sx={{
                flex: 2,
              }}
            >
              <NoDirectMessages />
            </Box>
          )}
        </Stack>
      </Stack>
    </Container>
  );
}

export default Messages;
