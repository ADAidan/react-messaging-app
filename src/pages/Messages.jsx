import * as React from "react";
import axios from "axios";
import Container from "@mui/material/Container";
import Grid from "@mui/material/Grid";
import Typography from "@mui/material/Typography";
import Paper from "@mui/material/Paper";
import { Box, IconButton, Stack } from "@mui/material";
import AddIcon from "@mui/icons-material/Add";
import Message from "../components/Message";
import MessageInput from "../components/MessageInput";
import ChatCard from "../components/ChatCard";
import NoMessages from "../components/NoMessages";
import NoDirectMessages from "../components/NoDirectMessages";
import UserContext from "../Context";

function Messages() {
  const user = React.useContext(UserContext);
  const messageContainerRef = React.useRef(null);
  const [directMessages, setDirectMessages] = React.useState([]);
  const [displayedMessages, setDisplayedMessages] = React.useState(null);
  const [selectedChat, setSelectedChat] = React.useState(null);

  React.useEffect(() => {
    if (!messageContainerRef.current) return;
    messageContainerRef.current.scrollTop =
      messageContainerRef.current.scrollHeight;
  }, [displayedMessages]);

  React.useEffect(() => {
    const getDirectMessages = async () => {
      try {
        const userID = sessionStorage.getItem("user");
        const response = await axios.get(
          `http://localhost:3000/users/${userID}/direct-messages`,
        );
        const directMessagesData = response.data;
        setDirectMessages(directMessagesData);
      } catch (error) {
        // eslint-disable-next-line no-console
        console.error("Axios Error:", error);
      }
    };
    getDirectMessages();
  }, []);

  React.useEffect(() => {
    if (!directMessages.length) return;
    setSelectedChat(1);
    const getDisplayedMessages = () => {
      const currentChat = directMessages.find(
        (chat) => chat.id === selectedChat,
      );
      if (!currentChat) return [];
      return currentChat.messages;
    };
    const messagesData = getDisplayedMessages();
    setDisplayedMessages(messagesData);
  }, [directMessages, selectedChat]);

  const handleClickAddChat = () => {
    // push a new chat to the user's directMessages
  };

  const handleMouseDownAdd = (e) => {
    e.preventDefault();
  };

  return (
    <Container
      maxWidth="lg"
      sx={{
        height: "100vh",
        my: 0,
      }}
    >
      <Grid container spacing={2}>
        <Grid item xs={4}>
          <Paper
            elevation={3}
            sx={{
              pt: 10,
              pb: 1,
              px: 2,
              my: 0,
              height: "100vh",
              overflowY: "auto",
              borderRadius: 0,
            }}
          >
            <Stack
              direction="row"
              spacing={2}
              sx={{
                justifyContent: "space-between",
                alignItems: "center",
              }}
            >
              <Typography variant="h5" component="h1" sx={{ mb: 2, px: 1 }}>
                Direct Message
              </Typography>
              <IconButton
                aria-label="add chat"
                onClick={handleClickAddChat}
                onMouseDown={handleMouseDownAdd}
              >
                <AddIcon />
              </IconButton>
            </Stack>
            {directMessages.length > 0 &&
              directMessages.map((chat) => (
                <ChatCard
                  key={chat.id}
                  chat={chat}
                  setSelectedChat={setSelectedChat}
                />
              ))}
          </Paper>
        </Grid>
        <Grid item xs={8}>
          {selectedChat ? (
            <Paper
              elevation={3}
              sx={{
                position: "relative",
                pt: 10,
                pb: 1,
                px: 3,
                my: 0,
                height: "100vh",
                borderRadius: 0,
              }}
            >
              <Box
                data-testid="messages-container"
                ref={messageContainerRef}
                sx={{
                  overflowY: "auto",
                  p: 1,
                  height: "calc(100vh - 150px)",
                }}
              >
                <Grid
                  data-testid={`chat-${selectedChat}`}
                  container
                  spacing={2}
                >
                  {displayedMessages ? (
                    displayedMessages.map((message) => (
                      <Message key={message.id} message={message} />
                    ))
                  ) : (
                    <NoMessages />
                  )}
                </Grid>
              </Box>
              <MessageInput
                setDisplayedMessages={setDisplayedMessages}
                username={user.username}
              />
            </Paper>
          ) : (
            <NoDirectMessages />
          )}
        </Grid>
      </Grid>
    </Container>
  );
}

export default Messages;
