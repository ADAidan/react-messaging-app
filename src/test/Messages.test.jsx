import * as React from "react";
import { render, screen, describe, expect, it } from "@testing-library/react";
import Messages from "../pages/Messages";

describe("Messages", () => {
  it("should render the Direct Message header", () => {
    render(<Messages />);
    const headerElement = screen.getByRole("heading", {
      name: /Direct Message/i,
    });
    expect(headerElement).toBeInTheDocument();

    const searchInput = screen.getByRole("textbox", { name: /Message/i });
    expect(searchInput).toBeInTheDocument();
  });
});
