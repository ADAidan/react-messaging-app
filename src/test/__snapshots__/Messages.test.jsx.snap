// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Messages > should render the Direct Message Component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-emql7a-MuiContainer-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 css-1k23dfp-MuiGrid-root"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-1tiru08-MuiPaper-root"
            >
              <div
                class="MuiStack-root css-1fdhh1r-MuiStack-root"
              >
                <h1
                  class="MuiTypography-root MuiTypography-h5 css-oahfe1-MuiTypography-root"
                >
                  Direct Message
                </h1>
                <button
                  aria-label="add chat"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="AddIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8 css-vok43g-MuiGrid-root"
          >
            <div
              class="MuiStack-root css-llo7ln-MuiStack-root"
              test-id="no-direct-messages"
            >
              <h3
                class="MuiTypography-root MuiTypography-h3 css-1wlq30w-MuiTypography-root"
              >
                No direct messages
              </h3>
              <p
                class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
              >
                Start a conversation by adding a new chat
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-emql7a-MuiContainer-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 css-1k23dfp-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-1tiru08-MuiPaper-root"
          >
            <div
              class="MuiStack-root css-1fdhh1r-MuiStack-root"
            >
              <h1
                class="MuiTypography-root MuiTypography-h5 css-oahfe1-MuiTypography-root"
              >
                Direct Message
              </h1>
              <button
                aria-label="add chat"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="AddIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8 css-vok43g-MuiGrid-root"
        >
          <div
            class="MuiStack-root css-llo7ln-MuiStack-root"
            test-id="no-direct-messages"
          >
            <h3
              class="MuiTypography-root MuiTypography-h3 css-1wlq30w-MuiTypography-root"
            >
              No direct messages
            </h3>
            <p
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              Start a conversation by adding a new chat
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
